### PR TITLE
feat: support for strict enum mapping

### LINKED ISSUE #325: Support for strict enum mapping
**The problem**
The major advantage of using Mapperly over other reflection based solutions like AutoMapper is that we now have mapping validation at build time. This works fine for most objects but not for enums, this is still done at run time when mapping a source object to a destination type. 

**The solution**
A feature that can be enabled for explicit enum mapping. When the feature is enabled Mapperly should enforce that all enum values in the destination enum can be created using the values of the source enum. Possibly using a ```MapperAttribute```. When it's not possible to create all enum values, the build should fail.

**Alternatives**
Writing unit tests and run them during the build step in the CI/CD pipeline to enforce integrity.

### PR DESCRIPTION
**Pull request by jvzijp**
Closes #325

### CODE REVIEW

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs [hunk 1]
```diff
@@ -83,6 +83,18 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
                 ctx.Target);
         }
 
+        var targetMembers = ctx.Target
+            .GetMembers()
+            .OfType<IFieldSymbol>();
+
+        if (enumMemberMappings.Count < targetMembers.Count())
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:90]: This does not really detect all unmapped items. Eg. if the source enum contains the values `Value1, Value2, Value3` and the target enum contains the values `Value10`, `Value20`, `Value30` comparing the count does not work. It should work ByValue, ByName and also take the case sensitivity settings into account.

File: src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs [hunk 2]
```diff
@@ -269,4 +269,13 @@ internal static class DiagnosticDescriptors
         DiagnosticCategories.Mapper,
         DiagnosticSeverity.Info,
         true);
+
+    public static readonly DiagnosticDescriptor NotAllEnumValuesMapped = new DiagnosticDescriptor(
+        "RMG034",
+        "Not all enum values in the target enum are mapped",
+        "Not all enum values in the target enum are mapped from {0} to {1}",
```
Comment (latonz) [inline src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:276]: Instead of one generic diagnostic, I think `Enum member {0} on {1} not found on source enum {2}` and `Enum member {0} on {1} not found on target enum {2}` diagnostics help a lot during troubleshooting. It would also align with how properties diagnostics are reported.

**Review by latonz — CHANGES_REQUESTED** [review]
Thank you for your contribution! I added my feedback 😊

File: src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs [hunk 3]
```diff
@@ -269,4 +269,13 @@ internal static class DiagnosticDescriptors
...
+        DiagnosticCategories.Mapper,
+        DiagnosticSeverity.Warning,
```
Comment (latonz) [inline src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:278]: To not introduce a breaking change, I think the default severity should be Info.

Comment (jvzijp) [reply src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs]: Agreed, fixed now.

File: src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs (hunk 2 above)
Comment (jvzijp) [reply src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs]: Agreed, fixed now.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs (hunk 1 above)
Comment (jvzijp) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs]: I changed the implementation to look for missing values in source & target enums. All variations "ByValue, ByName and case sensitivity" are taken into account.

**Comment by jvzijp** [comment]
> Thank you for your contribution! I added my feedback 😊 Could you also adjust the docs (`01-mapper.md > Strict property mappings` and `04-enum.mdx`)?

I updated `01-mapper.md > Strict property mappings` but I'm not sure what to add to `04-enum.mdx`, what did you have in mind?

**Review by latonz — CHANGES_REQUESTED** [review]
Thank you for the updates 😊 😊 
I added my feedback and also added a separate comment what I had in mind regarding `04-enum.mdx`.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs [hunk 4]
```diff
@@ -83,6 +107,20 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
...
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.NotAllSourceEnumValuesMapped, missingSourceMembers.First(), ctx.Source, ctx.Target);
+        }
+
+        var missingTargetMembers = targetFieldsByName.Select(item => item.Key).Except(enumMemberMappings.Select(mapping => mapping.Value));
+
+        if (missingTargetMembers.Any())
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.NotAllTargetEnumValuesMapped, missingTargetMembers.First(), ctx.Target, ctx.Source);
+        }
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:122]: A few comments on this:

- Your proposed solution replicates the mapping done earlier with it's own implementation. I think this diagnostic should be based on the actual mapping created. 
- My proposed solution does in fact work for ignoreCase=true since it works with the actual mappings created. 
- I like that your solution reports ALL missing members instead of only the first

[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:122]: You are totally correct, my bad! I'm also fine with your solution. Probably just change the any to a foreach.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs [hunk 5]
```diff
@@ -51,12 +52,35 @@ private static TypeMapping BuildCastMappingAndDiagnostic(MappingBuilderContext c
             ctx.Source.ToDisplayString(),
             ctx.Target.ToDisplayString()
         );
+        return BuildEnumToEnumCastMapping(ctx);
+    }
+
+    private static TypeMapping BuildEnumToEnumCastMapping(MappingBuilderContext ctx)
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:58]: The current name implies that it expects both the source and target member are of type Enum.

File: test/Riok.Mapperly.Tests/Mapping/EnumTest.cs [hunk 6]
```diff
@@ -116,7 +122,7 @@ public Task EnumToOtherEnumByNameWithoutOverlap()
             "enum E2 {D, E, F}"
         );
 
-        return TestHelper.VerifyGenerator(source);
+        return TestHelper.VerifyGenerator(source, TestHelperOptions.AllowInfoDiagnostics);
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply test/Riok.Mapperly.Tests/Mapping/EnumTest.cs]: My bad, removed it.

File: test/Riok.Mapperly.Tests/Mapping/QueryableProjectionEnumTest.cs [hunk 7]
```diff
@@ -48,6 +48,94 @@ TestSourceBuilderOptions.Default with
             );
     }
 
+    [Fact]
+    public void EnumToAnotherEnumByNameMissingTargetValueShouldDiagnostic()
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply test/Riok.Mapperly.Tests/Mapping/QueryableProjectionEnumTest.cs]: All tests are moved to `EnumTest.cs`

File: test/Riok.Mapperly.Tests/Mapping/QueryableProjectionEnumTest.cs [hunk 8]
```diff
@@ -48,6 +48,94 @@ TestSourceBuilderOptions.Default with
...
+    {
+        var source = TestSourceBuilder.Mapping(
+            "A",
+            "B",
+            TestSourceBuilderOptions.Default with
+            {
+                EnumMappingStrategy = EnumMappingStrategy.ByName
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply test/Riok.Mapperly.Tests/Mapping/QueryableProjectionEnumTest.cs]: Added some CaseInsensitive tests.

File: docs/docs/02-configuration/01-mapper.md [hunk 9]
```diff
@@ -91,3 +91,15 @@ set the following two EditorConfig settings (see also [analyzer diagnostics](./1
...
+To enforce strict enum mappings
+(all source enum values have to be mapped to a target enum value
+and all target enum values have to be mapped from a source enum value)
+set the following two EditorConfig settings (see also [analyzer diagnostics](./13-analyzer-diagnostics.mdx)):
+
+```editorconfig title=".editorconfig"
+[*.cs]
+dotnet_diagnostic.RMG034.severity = error # Unmapped target enum value
+dotnet_diagnostic.RMG035.severity = error # Unmapped source enum value
+```
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply docs/docs/02-configuration/01-mapper.md]: Done.

File: src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs [hunk 10]
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
         DiagnosticSeverity.Info,
         true
     );
+
+    public static readonly DiagnosticDescriptor NotAllTargetEnumValuesMapped = new DiagnosticDescriptor(
+        "RMG034",
+        "Not all enum values in the target enum are mapped",
+        "Enum member {0} on {1} not found on source enum {2}",
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs]: Fixed.

[the remark this answers is not in the document]
Comment (jvzijp) [reply src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs]: I agree. Fixed.

File: src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs [hunk 11]
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
         DiagnosticSeverity.Info,
         true
     );
+
+    public static readonly DiagnosticDescriptor NotAllTargetEnumValuesMapped = new DiagnosticDescriptor(
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs]: Done.

File: src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs [hunk 12]
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
...
+        "RMG034",
+        "Not all enum values in the target enum are mapped",
+        "Enum member {0} on {1} not found on source enum {2}",
+        DiagnosticCategories.Mapper,
+        DiagnosticSeverity.Info,
+        true
+    );
+
+    public static readonly DiagnosticDescriptor NotAllSourceEnumValuesMapped = new DiagnosticDescriptor(
```
[the remark this answers is not in the document]
Comment (jvzijp) [reply src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs]: Done.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs [hunk 13]
```diff
@@ -78,11 +97,31 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
             .Where(x => x.Target != null)
             .ToDictionary(x => x.Source.Name, x => x.Target!.Name);
 
-        if (enumMemberMappings.Count == 0)
+        if (!enumMemberMappings.Any())
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:100]: Is this change needed? Isn't the count check faster?

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs [hunk 14]
```diff
@@ -78,11 +97,31 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
...
         {
             ctx.ReportDiagnostic(DiagnosticDescriptors.EnumNameMappingNoOverlappingValuesFound, ctx.Source, ctx.Target);
         }
 
+        var missingSourceMembers = sourceFieldsByName.Where(field => !enumMemberMappings.ContainsKey(field.Key));
+        foreach (var member in missingSourceMembers)
+            ctx.ReportDiagnostic(
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:107]: This code base uses braces for foreach and ifs, except if it is only a return/throw statement.

File: test/Riok.Mapperly.Tests/Mapping/EnumTest.cs [hunk 15]
```diff
@@ -332,4 +338,61 @@ public void EnumToStringMappingDisabledShouldUseToStringMapping()
...
+            {
+                EnumMappingStrategy = enumMappingStrategy
+            },
+            "class A { public C Value { get; set; } }",
+            "class B { public D Value { get; set; } }",
+            $"enum C {{ {sourceEnumValues} }}",
+            $"enum D {{ {targetEnumValues} }}"
+        );
+
+        return TestHelper.VerifyGenerator(source, null, enumMappingStrategy, sourceEnumValues, targetEnumValues);
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Mapping/EnumTest.cs:370]: Instead of using all values as parameters, I'd use an additional test parameter describing the test case, what it actually tests (eg. `by name with missing source value`).
Be reminded to remove unused snapshot files after updating the snapshot name.

File: test/Riok.Mapperly.Tests/TestHelper.cs [hunk 16]
```diff
@@ -7,10 +7,15 @@ namespace Riok.Mapperly.Tests;
...
 {
-    public static Task VerifyGenerator(string source, TestHelperOptions? options = null)
+    public static Task<VerifyResult> VerifyGenerator(string source, TestHelperOptions? options = null, params object?[] args)
     {
         var driver = Generate(source, options);
-        return Verify(driver).ToTask();
+        var verify = Verify(driver);
+
+        if (args.Any())
+            verify.UseParameters(args);
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/TestHelper.cs:16]: This code base uses braces for ifs except if it is a simple return/throw statement.

**Review by latonz — CHANGES_REQUESTED** [review]
Thanks for the updates.
There are still a few linter issues...

File: test/Riok.Mapperly.Tests/TestHelper.cs (hunk 16 above)
Comment (jvzijp) [reply test/Riok.Mapperly.Tests/TestHelper.cs:17]: Fixed.

File: test/Riok.Mapperly.Tests/Mapping/EnumTest.cs (hunk 15 above)
Comment (jvzijp) [reply test/Riok.Mapperly.Tests/Mapping/EnumTest.cs]: Fixed.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs (hunk 14 above)
Comment (jvzijp) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:112]: Fixed.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs (hunk 13 above)
Comment (jvzijp) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs]: Is indeed faster than Any(), it's just my personal preference. I've no issue with changing it back :)

**Review by latonz — DISMISSED** [review]

**Comment by jvzijp** [comment]
How do I fix the conventional commits check?

**Review by latonz — APPROVED** [review]

**Comment by latonz** [comment]
@jvzijp no need to make this check green since it is not a required one. It's just a reminder to adjust the squashed commit message when merging the PR as needed.
Thank you for the contribution 😊

**Comment by jvzijp** [comment]
> 

Ah I see. Great :) My pleasure, it was a great learning experience.
